File: support/src/python/statistics/validation_utils.py
"""
validation_utils - Validation framework utilities.

Provides structured validation for experiment results.
"""

from dataclasses import dataclass, field
from typing import List, Any, Dict, Optional
import operator


@dataclass
class ValidationResult:
    """Container for validation results."""
    passed: bool
    messages: List[str] = field(default_factory=list)
    
    def __bool__(self) -> bool:
        return self.passed
    
    @classmethod
    def success(cls, message: str) -> 'ValidationResult':
        """Create a successful validation result."""
        return cls(passed=True, messages=[message])
    
    @classmethod
    def failure(cls, message: str) -> 'ValidationResult':
        """Create a failed validation result."""
        return cls(passed=False, messages=[message])


COMPARISON_OPS = {
    '>=': operator.ge,
    '<=': operator.le,
    '>': operator.gt,
    '<': operator.lt,
    '==': operator.eq,
}


def validate_threshold(
    value: float,
    threshold: float,
    comparison: str = '>=',
    name: str = 'Value'
) -> ValidationResult:
    """
    Validate a value against a threshold.
    
    Args:
        value: The value to check
        threshold: The threshold to compare against
        comparison: Comparison operator ('>=', '<=', '>', '<', '==')
        name: Name for the validation message
        
    Returns:
        ValidationResult
    """
    if comparison not in COMPARISON_OPS:
        raise ValueError(f"Unknown comparison operator: {comparison}")
    
    op = COMPARISON_OPS[comparison]
    result = op(value, threshold)
    
    status = 'PASS' if result else 'FAIL'
    message = f"{status}: {name} = {value:.4f} {comparison} {threshold}"
    
    return ValidationResult(passed=result, messages=[message])


def validate_metrics(
    stats: Any,
    config: Dict[str, Any]
) -> ValidationResult:
    """
    Validate statistics against configuration thresholds.
    
    Checks validation_min_vaf and validation_max_rmse from config.
    
    Args:
        stats: Statistics object with .vaf and .rmse attributes
        config: Configuration dictionary
        
    Returns:
        ValidationResult
    """
    min_vaf = config.get('validation_min_vaf', float('-inf'))
    max_rmse = config.get('validation_max_rmse', float('inf'))
    
    validations = [
        validate_threshold(stats.vaf, min_vaf, '>=', 'VAF (%)'),
        validate_threshold(stats.rmse, max_rmse, '<=', 'RMSE')
    ]
    
    return validate_all(validations)


def validate_all(validations: List[ValidationResult]) -> ValidationResult:
    """
    Combine multiple validation results.
    
    Overall passes only if all pass.
    
    Args:
        validations: List of ValidationResult objects
        
    Returns:
        Combined ValidationResult
    """
    all_passed = all(v.passed for v in validations)
    all_messages = []
    for v in validations:
        all_messages.extend(v.messages)
    
    return ValidationResult(passed=all_passed, messages=all_messages)

File: support/src/julia/statistics/ValidationUtils.jl
"""
ValidationUtils - Validation framework utilities.

Provides structured validation for experiment results.

# Exports
- `ValidationResult`: Struct for validation outcomes
- `validate_threshold`: Check value against threshold
- `validate_all`: Run multiple validations
"""
module ValidationUtils

export ValidationResult, validate_threshold, validate_all

"""
    struct ValidationResult

Container for validation results.

# Fields
- `passed::Bool`: Whether validation passed
- `messages::Vector{String}`: Validation messages
"""
struct ValidationResult
    passed::Bool
    messages::Vector{String}
end

"""
    ValidationResult(passed::Bool, message::String)

Create a ValidationResult with a single message.
"""
ValidationResult(passed::Bool, message::String) = ValidationResult(passed, [message])

"""
    passed(result::ValidationResult) -> Bool

Check if validation passed.
"""
passed(result::ValidationResult) = result.passed

"""
    validate_threshold(value, threshold; comparison=:>=, name="Value") -> ValidationResult

Validate a value against a threshold.

# Arguments
- `value`: The value to check
- `threshold`: The threshold to compare against
- `comparison`: Comparison operator (:>=, :<=, :>, :<, :==)
- `name`: Name for the validation message
"""
function validate_threshold(value, threshold; comparison=:>=, name::String="Value")::ValidationResult
    result = if comparison == :>=
        value >= threshold
    elseif comparison == :<=
        value <= threshold
    elseif comparison == :>
        value > threshold
    elseif comparison == :<
        value < threshold
    elseif comparison == :(==)
        value == threshold
    else
        error("Unknown comparison operator: $comparison")
    end
    
    status = result ? "PASS" : "FAIL"
    message = "$status: $name = $(round(value, digits=4)) $comparison $threshold"
    
    return ValidationResult(result, message)
end

"""
    validate_all(validations::Vector{ValidationResult}) -> ValidationResult

Combine multiple validation results. Overall passes only if all pass.
"""
function validate_all(validations::Vector{ValidationResult})::ValidationResult
    all_passed = all(v -> v.passed, validations)
    all_messages = reduce(vcat, [v.messages for v in validations])
    return ValidationResult(all_passed, all_messages)
end

end # module
